controllers/metal3.io/hostfirmwarecomponents_controller.go

```diff
@@ -0,0 +1,319 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/go-logr/logr"
+	metal3api "github.com/metal3-io/baremetal-operator/apis/metal3.io/v1alpha1"
+	"github.com/metal3-io/baremetal-operator/pkg/provisioner"
+)
+
+// HostFirmwareComponentsReconciler reconciles a HostFirmwareComponents object.
+type HostFirmwareComponentsReconciler struct {
+	client.Client
+	Log                logr.Logger
+	ProvisionerFactory provisioner.Factory
+}
+
+// rhfcInfo is used to simplify the pass or arguments.
+type rhfcInfo struct {
+	ctx    context.Context
+	log    logr.Logger
+	hfc    *metal3api.HostFirmwareComponents
+	bmh    *metal3api.BareMetalHost
+	events []corev1.Event
+}
+
+type conditionReasonHFC string
+
+const (
+	reasonInvalidComponent conditionReasonHFC = "InvalidComponent"
+	reasonValidComponent   conditionReasonHFC = "OK"
+)
+
+func (info *rhfcInfo) publishEvent(reason, message string) {
+	t := metav1.Now()
+	hfcEvent := corev1.Event{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: reason + "-",
+			Namespace:    info.hfc.ObjectMeta.Namespace,
+		},
+		InvolvedObject: corev1.ObjectReference{
+			Kind:       "HostFirmwareComponents",
```

Comment:
nit: there is probably a constant for this value somewhere in `apis`

Reply:
ack, I will check it

Reply:
I've looked in what is present in apis/metal3.io/v1alpha1, I don't have any constant for "HostFirmwareComponents" (I also couldn't find any for other CRDs, like BareMetalHost etc..)

Reply:
Yeah, it's less trivial than I thought. You probably can get the kind from the object itself, like `info.hfc.Kind`.